app/components/Item.js

```diff
@@ -1,78 +1,45 @@
-import React, { Component, PropTypes } from 'react'
-import { DragSource, DropTarget } from 'react-dnd'
+import React from 'react'
 import Tree from './Tree'
-
-const source = {
-  beginDrag(props) {
-    return {
-      id: props.id,
-      parent: props.parent,
-      items: props.item.children
-    }
-  },
-
-  isDragging(props, monitor) {
-    return props.id == monitor.getItem().id
-  }
-}
-
-const target = {
-  canDrop() {
-    return false
-  },
-
-  hover(props, monitor) {
-    const {id: draggedId} = monitor.getItem()
-    const {id: overId} = props
-
-    if (draggedId == overId || draggedId == props.parent) return
-    if (!monitor.isOver({shallow: true})) return
-
-    props.move(draggedId, overId, props.parent)
-  }
-}
-
-@DropTarget('ITEM', target, connect => ({
-  connectDropTarget: connect.dropTarget()
-}))
-@DragSource('ITEM', source, (connect, monitor) => ({
-  connectDragSource: connect.dragSource(),
-  connectDragPreview: connect.dragPreview(),
-  isDragging: monitor.isDragging()
-}))
-export default class Item extends Component {
-  static propTypes = {
-    id     : PropTypes.any.isRequired,
-    parent : PropTypes.any,
-    item   : PropTypes.object,
-    move   : PropTypes.func,
-    find   : PropTypes.func
-  };
-
-  render() {
-    const {
-      connectDropTarget, connectDragPreview, connectDragSource,
-      item: {id, title, children}, parent, move, find
-    } = this.props
-
-    return connectDropTarget(connectDragPreview(
-      <div>
-        {connectDragSource(
-          <div style={{
-            background: 'white',
-            border: '1px solid #ccc',
-            padding: '1em',
-            marginBottom: -1
-          }}
-          >{title}</div>
-        )}
-        <Tree
-          parent={id}
-          items={children}
-          move={move}
-          find={find}
-        />
+import { useDrag, useDrop } from 'react-dnd'
+
+const Item = ({ item, parentId, moveItem }) => {
+  const [{}, dropRef] = useDrop({
+    accept: 'item',
+    canDrop: false,
+    hover: (draggedItem, monitor) => {
+      if (draggedItem.id === item.id || draggedItem.id === parentId) return
+      if (!monitor.isOver({ shallow: true })) return
+
+      moveItem(draggedItem.id, item.id, parentId)
+    },
+  })
+
+  const [{}, dragRef] = useDrag({
```

Comment:
pull out a preview ref here so we can set the previewing component: `const [{}, dragRef, previewRef]`

Reply:
Interesting, so there exists the ability to show custom preview components? Nice!

Reply:
Also, maybe we can remove the ` {}` here, because `eslint` is complaining on those lines.
So it will become `const [, dragRef, previewRef] = ...`

Reply:
yes, that would be great!

Reply:
The official `react-dnd` docs didn't seem to have any info about `previewRef` thing in the `useDrag`...
But yeah, that's awesome that we can specify that!

Reply:
🤷 yeah, i'm not sure why the docs are lacking this. I found a codesandbox example using it.